plat/kvm/arm/setup.c

```diff
@@ -205,6 +207,76 @@ static void _dtb_get_cmdline(char *cmdline, size_t maxlen)
 	uk_pr_info("No command line found\n");
 }
 
+#ifdef CONFIG_PAGING
+
+int _init_paging(void)
+{
+	int rc;
+	uint64_t start;
+	uint64_t len;
+	unsigned long frames;
+	__sz free_memory, res_memory;
+
+	/* Assign all available memory beyond the boot stack
+	 * to the frame allocator.
+	 */
+	start = ALIGN_UP(__END + __STACK_SIZE, PAGE_SIZE);
+	len   = _libkvmplat_cfg.bstack.end - start;
```

Comment:
There seems to be something wrong here, or I just don't get it 🤔 In lines 160+ the boot stack is positioned at the end of the available memory, but here you set things up for locating it right behind the image. The latter is also in line with the comment before the configuration of `heap.start`. But then in lines 266+ you configure the boot stack to be behind the heap again.

Reply:
Okay, so during entry, the stack is set up to follow the kernel image (entry64.S:80). Later, after the device tree is parsed and the amount of available memory is determined, the stack is positioned to the end of the image (setup.c:160). Without paging, we would later switch to the new stack (setup.c:341).

When paging is enabled, `_init_paging()` is called before we migrate to the new stack. The calculation above, is therefore based on the current one, ie the early stack. I think the confusion is caused by KVM using the term "boot stack" ambiguously. The way I see it is that the boot stack is the early stack. The one we later switch to is just "the stack", as it is used for all runtime - not just boot. If we agree on that, I can also add a patch to rename `_libkvmplat_cfg.bstack` to `_libkvmplat_cfg.stack`, along with other references. I wanted to change that for a while now 😈

Reply:
This makes sense now 😃 We should have the following layout over time then:
![image](https://user-images.githubusercontent.com/8928900/184336522-54bd456a-002f-4e80-888c-282de9e582a3.png)

So we are "wasting" the true boot stack that is accounted for via `STACK_SIZE`, right?